## Exercise 1 

### a)


Desde que $v$=$(\textbf{v}_{1},\textbf{v}_{2},\textbf{v}_{3})$ é uma base de um espaço vetorial $V$, qualquer elemento $m \in V$ é descrito como uma combinação linear de elemtos da base, ou seja:

$$\textbf{m} = x\;\textbf{v}_{1} + y\;\textbf{v}_{2} + z\;\textbf{v}_{3} \;\; , \;\;\ x,y,z \; \in \; \Re$$

$$\Leftrightarrow \textbf{m} = (x,y,z)_{v}$$
$$\Rightarrow f(\textbf{m}) = x\;f(\textbf{v}_{1}) + y\;f(\textbf{v}_{2})
+ z\;f(\textbf{v}_{3})$$


$$\Rightarrow f(\textbf{m}) = x  
    \begin{bmatrix}%array with vert lines
        4 \\
        0 \\
        -2\;\;
    \end{bmatrix} + y 
    \begin{bmatrix}%array with vert lines
        -2\;\; \\
        0 \\
        1
    \end{bmatrix} + z 
    \begin{bmatrix}%array with vert lines
        2 \\
        1 \\
       -1\;\;
    \end{bmatrix}$$

$$\Leftrightarrow f(\textbf{m}) = 
\begin{bmatrix}
   4 &-2\;\;&2 \\
   0 &0 &1 \\
   -2\;\; &1 &-1\;\;
\end{bmatrix}
\cdot
\begin{bmatrix}
    x \\
    y \\
    z \\
\end{bmatrix}$$

$$\Leftrightarrow f(\textbf{m}) =
\begin{bmatrix}
   4 &-2\;\;&2 \\
   0 &0 &1 \\
   -2\;\; &1 &-1\;\;
\end{bmatrix}
\cdot
\textbf{m}$$

Portanto, a matriz correspondente à transformação linear $f : V \rightarrow \Re^{3}$ é :

$$A_{v}=
\begin{bmatrix}
   4 &-2\;\;&2 \\
   0 &0 &1 \\
   -2\;\; &1 &-1\;\;
\end{bmatrix}$$

Logo,
$$f((17,-15,16))=A_{v} \cdot 
\begin{bmatrix}
    17 \\
    -15\;\; \\
    16 \\
\end{bmatrix}$$

In [9]:
import numpy as np

A=np.array([[4 ,-2 ,2],
           [0, 0, 1],
           [-2, 1, -1]])
m= np.array([17,-15,16])


In [11]:
np.dot(A,m)

array([130,  16, -65])

Logo,
$$f((17,-15,16))=
\begin{bmatrix}
    130 \\
    16\\
    -65\;\; \\
\end{bmatrix}$$

### b)

Dada a matriz $A$ , o kernel é definido por $\textbf{u}$ tal que $A\cdot \textbf{u}=0$

escalonando a matriz A:



In [13]:
def pivotamento(A):
    """A função recebe uma matriz A (n x m) e
    retorna a matriz zerada na diagonal inferior
    esquerda."""
    for i in range(len(A) - 1):
        for j in range(i + 1,len(A)):
            n1 = A[j][i]
            if n1 == 0:
                for k in range(i,len(A)):
                    if A[k][i] != 0:
                        swap(A, i, k)
                        n1 = A[j][i]
            n2 = A[i][i]
            if n2 == 0:
                continue
            div = n1 / n2
            A[j] = sumVector(multVector(A[i], -div), A[j])
    return A

def multVector(v, x):
    newV = []
    for i in range(len(v)):
        newV += [v[i] * x]
    return newV

def sumVector(v1, v2):
    newV = []
    for i in range(len(v1)):
        newV += [v1[i] + v2[i]]
    return newV

def swap(A, i, j):
    newI = []
    newJ = []
    for index in range(len(A) ):
        newI += [A[j][index]]
        newJ += [A[i][index]]
    A[i] = newI
    A[j] = newJ

def printMatrix(A):
    for i in A:
        print ("|\t", end="")
        for j in i:
            print ("%.1f" %(j), "\t", end="")
        print ("  |")
        print("\n")

#Matriz A
A = [[4 ,-2 ,2], [0, 0, 1], [-2, 1, -1]]

printMatrix(A)
print("\n\n")
B=pivotamento(A)
print("")
printMatrix(B)

|	4.0 	-2.0 	2.0 	  |


|	0.0 	0.0 	1.0 	  |


|	-2.0 	1.0 	-1.0 	  |






|	-2.0 	1.0 	-1.0 	  |


|	0.0 	0.0 	1.0 	  |


|	0.0 	0.0 	0.0 	  |




Assim, é equivalente a resolver :

$$\Leftrightarrow
\begin{bmatrix}
    -2\;\; &1 &-1\;\; \\
    0 &0 &1 \\
    0 &0 &0
\end{bmatrix} \cdot \textbf{u} = 0$$

$$\Leftrightarrow \begin{bmatrix}
    -2\;\; \\
    0 \\
    0
\end{bmatrix} \cdot x +
\begin{bmatrix}
    1 \\
    0 \\
    0
\end{bmatrix} \cdot y +
\begin{bmatrix}
    -1\;\; \\
    1 \\
    0
\end{bmatrix} \cdot z = 0$$

Logo, $z=0$ , $y=2\;x\;\;$ e , então:

$$\textbf{u}=x\cdot
\begin{bmatrix}
    2 \\
    1 \\
    0
\end{bmatrix} \;\;\;\;\;\;\; ,\forall x \in \Re$$

portanto, 

$$\begin{bmatrix}
    2 \\
    1 \\
    0
\end{bmatrix} \;\;, \;\;
\begin{bmatrix}
    4 \\
    2 \\
    0
\end{bmatrix}$$ são dois vetores pertencentes a $ker(f)$

Logo, $dim(Ker(f))=1$ e ,pelo Teorema Núcleo-Imagem, $dim(range(f(V)))=2$

## B)  - REMAKE


In [46]:
import sympy as sp





In [47]:


F=sp.Matrix([[4.0,-2.0,2.0],
            [0.0,0.0,1.0],
            [-2.0,1.0,-1.0]
            ])
#F=F.transpose()

In [48]:
F.rref()

(Matrix([
 [  1, -0.5, 0],
 [0.0,  0.0, 1],
 [  0,    0, 0]]), (0, 2))

In [49]:
F.nullspace()

[Matrix([
 [0.5],
 [  1],
 [  0]])]

#### PODE-SE DETERMINAR a matriz de g pela sua transformação inversa, ou seja:

$$g^{-1}(g(\textbf{u}_{i}))=g^{-1}(\textbf{v}_{i})$$

Sendo:

$$g=(g^{-1})^{-1}$$

In [53]:
#Transformação Inversa
G_inv=sp.Matrix([[1.0,0.0,1.0],
            [0.0,1.0,0.0],
            [1/2,0.0,-1/2]
            ])
G_inv=G_inv.transpose()

In [56]:
G_inv

Matrix([
[1.0, 0.0,  0.5],
[0.0, 1.0,  0.0],
[1.0, 0.0, -0.5]])

In [54]:
G=G_inv**-1

In [55]:
G

Matrix([
[0.5,   0,  0.5],
[  0, 1.0,    0],
[1.0,   0, -1.0]])

In [58]:
G.columnspace() #list

[Matrix([
 [0.5],
 [  0],
 [1.0]]), Matrix([
 [  0],
 [1.0],
 [  0]]), Matrix([
 [ 0.5],
 [   0],
 [-1.0]])]

In [59]:
A=G.columnspace()
len(A) # Legnth of A / DIM(A)

3

In [57]:
h=G*F
h

Matrix([
[1.0, -0.5, 0.5],
[  0,    0, 1.0],
[6.0, -3.0, 3.0]])

In [60]:
h.eigenvals()

{3: 1, 1: 1, 0: 1}

In [62]:
A=h.eigenvects()
A

[(0, 1, [Matrix([
   [0.5],
   [1.0],
   [  0]])]), (1.00000000000000, 1, [Matrix([
   [0.166666666666667],
   [              1.0],
   [              1.0]])]), (3.00000000000000, 1, [Matrix([
   [0.166666666666667],
   [0.333333333333333],
   [              1.0]])])]

In [65]:
A[1]

(1.00000000000000, 1, [Matrix([
  [0.166666666666667],
  [              1.0],
  [              1.0]])])

In [67]:
sp.nsimplify(0.166666666666667)

1/6

### c)

$\{(1,0,1)\;,\;(0,1,0)\;,\;(\frac{1}{2},0,-\frac{1}{2} \}$ É L.I., sendo uma base do $\Re^{3}$ conforme o escalonamento:

In [14]:
A = [[1,0,1], [0, 1, 0], [1/2, 0, -1/2]]

printMatrix(A)
print("\n\n")
B=pivotamento(A)
print("")
printMatrix(B)

|	1.0 	0.0 	1.0 	  |


|	0.0 	1.0 	0.0 	  |


|	0.5 	0.0 	-0.5 	  |






|	0.5 	0.0 	-0.5 	  |


|	0.0 	1.0 	0.0 	  |


|	0.0 	0.0 	2.0 	  |




A matriz acima é a forma escalonada, portanto, a transformação tem núcleo de dimensão nula, ou seja $ker(g)=\{0\}$ . Pelo Teorema Núcleo-Imagem, $dim(range(g(\Re^{3})))=3$

### d)

In [41]:
A = [[1.0,0.0,1.0], [0.0, 1.0, 0.0], [1/2, 0.0, -1/2]]
B=np.transpose(A)
A=np.array([[4.0 ,-2.0 ,2.0],
           [0.0, 0.0, 1.0],
           [-2.0, 1.0, -1.0]])
print(B)

[[ 1.   0.   0.5]
 [ 0.   1.   0. ]
 [ 1.   0.  -0.5]]


Obteve-se acima a matriz $B$ correspondente à transformação linear $g(\Re^{3})$ por meio da matriz de:



$$f(\textbf{m}) =
\begin{bmatrix}
   4 &-2\;\;&2 \\
   0 &0 &1 \\
   -2\;\; &1 &-1\;\;
\end{bmatrix}
\cdot
\textbf{m}$$

$$g(f(\textbf{m}))=
\begin{bmatrix}
   1 &0 &\frac{1}{2} \\
   0 &0 &0 \\
   1 &0 &-\frac{1}{2}\;\;
\end{bmatrix}
\cdot
\Bigg(\begin{bmatrix}
   4 &-2\;\;&2 \\
   0 &0 &1 \\
   -2\;\; &1 &-1\;\;
\end{bmatrix}
\cdot
\textbf{m}\Bigg)
$$

Pela propriedade associativa, 
$$g(f(\textbf{m}))=
\Bigg(
\begin{bmatrix}
   1 &0 &\frac{1}{2} \\
   0 &0 &0 \\
   1 &0 &-\frac{1}{2}\;\;
\end{bmatrix}
\cdot
\begin{bmatrix}
   4 &-2\;\;&2 \\
   0 &0 &1 \\
   -2\;\; &1 &-1\;\;
\end{bmatrix}\Bigg)
\cdot
\textbf{m}
$$

finalmente,


In [42]:
M=np.dot(B,A)
print(M)

[[ 3.  -1.5  1.5]
 [ 0.   0.   1. ]
 [ 5.  -2.5  2.5]]


$$g(f(\textbf{m}))=
\begin{bmatrix}
   3 &-\frac{3}{2}\;\; &\frac{3}{2} \\
   0 &0 &1 \\
   5 &-\frac{5}{2} &-\frac{5}{2}\;\;
\end{bmatrix}
\cdot
\textbf{m}
$$

$$\Leftrightarrow h(\textbf{m}) = M_{3x3} \cdot \textbf{m} , \;\; \textbf{m} \in \textbf{V}$$

Portanto, $h(\lambda \textbf{m})=M \cdot (\lambda \;\textbf{m}) = \lambda(M\;\cdot\textbf{m})= \lambda h(\textbf{m})$ , pela propriedade associativa das matrizes.

Também,  $$h(\textbf{m} + \textbf{n}) =M_{3x3} \cdot(\textbf{m} + \textbf{n}) =M_{3x3} \cdot \textbf{m} +M_{3x3} \cdot \textbf{n} = h(\textbf{m}) + h(\textbf{n}) $$ pela propriedade distributiva das matrizes.

Resta descobrir  um fixvector  $w$ :

$$M\cdot \textbf{m} = \lambda \cdot \textbf{m} $$

$$\Leftrightarrow(M-\lambda I)\cdot \textbf{m} = 0$$

Basta descobrir os autovetores de $M$,  por meio do polinômio característico:

$det(M-\lambda I) = 0$


In [44]:
np.linalg.eig(M)

(array([5.00000000e+00, 6.02970065e-17, 5.00000000e-01]),
 array([[-5.07092553e-01, -4.47213595e-01, -2.59160528e-01],
        [-1.69030851e-01, -8.94427191e-01, -8.63868426e-01],
        [-8.45154255e-01, -1.84549588e-16, -4.31934213e-01]]))

escolhem-se 2 autovalores : $5$ e $\frac{1}{2}$

assim, podem-se encontrar elementos $\textbf{w}_{1}$ e $\textbf{w}_{2}$ tais que :

$$(M-5I)\textbf{w}_{1}=0$$

e 
$$(M-\frac{1}{2}I)\textbf{w}_{2}=0$$


In [48]:
I=np.array([[1.0,0.0,0.0],
            [0.0,1.0,0.0],
            [0.0,0.0,1.0]])
D=pivotamento(M-5*I)
print(D)

print("")

E=pivotamento(M-I/2)

print(E)

[[ 5.  -2.5 -2.5]
 [ 0.  -5.   1. ]
 [ 0.   0.   0. ]]

[[ 5.  -2.5  2. ]
 [ 0.  -0.5  1. ]
 [ 0.   0.   0. ]]


Logo,
$$\begin{bmatrix}
   5 &-\frac{5}{2}\;\; &-\frac{5}{2}\;\; \\
   0 &-5\;\; &1 \\
   0 &0 &0
\end{bmatrix}
\cdot
\textbf{w}_{1}=0$$

$$\begin{bmatrix}
   5 &-\frac{5}{2}\;\; &2 \\
   0 &-\frac{1}{2}\;\; &1 \\
   0 &0 &0
\end{bmatrix}
\cdot
\textbf{w}_{2}=0$$

$$\textbf{w}_{1}=\begin{bmatrix}
    \frac{3z}{5} \\
    \frac{z}{5} \\
    z
\end{bmatrix} = z
\begin{bmatrix}
    \frac{3}{5} \\
    \frac{1}{5} \\
    1
\end{bmatrix}$$

$$\textbf{w}_{2}=\begin{bmatrix}
    \frac{3z}{5} \\
    2z \\
    z
\end{bmatrix} = z
\begin{bmatrix}
    \frac{3}{5} \\
    2 \\
    1
\end{bmatrix},\\
\forall z \in \Re$$

## Exercise 2

### a)

Seja $\textbf{E}_{1}=
\begin{bmatrix}
    1 &0\\
    0 &0
\end{bmatrix}$ , 
$\textbf{E}_{2}=
\begin{bmatrix}
    0 &1\\
    0 &0
\end{bmatrix}$,
$\textbf{E}_{3}=
\begin{bmatrix}
    0 &0\\
    1 &0
\end{bmatrix}$,
$\textbf{E}_{4}=
\begin{bmatrix}
    0 &0\\
    0 &1
\end{bmatrix}$, então
$(\textbf{E}_{1},\textbf{E}_{2}, \textbf{E}_{3}, \textbf{E}_{4})$ é base canônica $E$ para $\Re^{2x2}$





Dessa forma, na base $E$:

$\textbf{A}_{1}=(1,-1,1,2),
\textbf{A}_{2}=(1,-1,2,1),
\textbf{A}_{3}=(1,-1,2,2)$

Dessa forma,

$$\begin{bmatrix}
1 &-1 &1 &2\\
1 &-1 &2 &1\\
1 &-1 &2 &2
\end{bmatrix} \sim M$$




In [46]:
import sympy as sp



In [61]:
M=sp.Matrix([[1, -1, 1, 2],
          [1, -1, 2, 1],
          [1, -1, 2, 2]])

print(M.rref())

(Matrix([
[1, -1, 0, 0],
[0,  0, 1, 0],
[0,  0, 0, 1]]), (0, 2, 3))


In [62]:
print(C)

[[ 1. -1.  1.  2.]
 [ 0.  1.  2. -4.]
 [ 0.  0.  1.  0.]
 [ 0.  0.  0.  1.]]


$$\begin{bmatrix}
1 &-1 &1 &2\\
1 &-1 &2 &1\\
1 &-1 &2 &2
\end{bmatrix} \sim
\begin{bmatrix}
1 &-1 &1 &2\\
0 &0 &1 &-1\\
0 &0 &1 &0
\end{bmatrix} = M$$

Portanto, constatamos que sua Echelon form não possui linha identicamente nula. Logo, o conjunto $a$ é LI, sendo, portanto, uma base do subspaço de $span\{a\}$

### b)

escrevendo-se as matrizes $\textbf{B}$ na base $E$, se um vetor for combinação linear de um conjunto LD, então surge uma linha nula na matriz cujas linhas são esses vetores.

Testando-se $\textbf{B}_{1}$

In [56]:
H=sp.Matrix([[1.0,-1.0,1.0,2.0],[1.0,-1.0,2.0,1.0],[1.0,-1.0,2.0,2.0],[2.0,-1.0,3.0,0.0]])
print(H.rref())

(Matrix([
[1, 0, 0, 0],
[0, 1, 0, 0],
[0, 0, 1, 0],
[0, 0, 0, 1]]), (0, 1, 2, 3))


Ao escalonar a matriz, percebe-se que o conjunto $(\textbf{A}_{1},\textbf{A}_{2},\textbf{A}_{3},\textbf{B}_{1})$ é LI

Testando-se $\textbf{B}_{2}$

In [68]:
H=sp.Matrix([[1.0,-1.0,1.0,2.0],
            [1.0,-1.0,2.0,1.0],
            [1.0,-1.0,2.0,2.0],
            [1.0,-1.0,1.0,1.0]])
print(H.rref())
#C=pivotamento(H)
#print(C)

(Matrix([
[1, -1.0, 0, 0],
[0,    0, 1, 0],
[0,    0, 0, 1],
[0,    0, 0, 0]]), (0, 2, 3))


Ao escalonar a matriz, percebe-se que o conjunto $(\textbf{A}_{1},\textbf{A}_{2},\textbf{A}_{3},\textbf{B}_{2})$ é LD.

Testando-se $\textbf{B}_{3}$

In [58]:
H=sp.Matrix([[1.0,-1.0,1.0,2.0],
            [1.0,-1.0,2.0,1.0],
            [1.0,-1.0,2.0,2.0],
            [2.0,-1.0,4.0,0.0]])
print(H.rref())

(Matrix([
[1, 0, 0, 0],
[0, 1, 0, 0],
[0, 0, 1, 0],
[0, 0, 0, 1]]), (0, 1, 2, 3))


Portanto, somente $B_{2} \in U$

Pode-se perceber, pelas operações linhas que anularam a última linha que:

    
$$\textbf{B}_{2}= 1\textbf{A}_{1}-2\textbf{A}_{2} +1\textbf{A}_{3}$$ 

Logo,
$$\textbf{B}_{2}=(1,-2,1)\begin{bmatrix}
\textbf{A}_{1}\\
\textbf{A}_{2}\\
\textbf{A}_{3}\\
\end{bmatrix}$$


### c)


É possível notar que 
$$\textbf{B}_{3}-\textbf{B}_{1}=\begin{bmatrix}
0 &0\\
1 &0
\end{bmatrix} =\textbf{A}_{3}-\textbf{A}_{1} $$

Portanto, a base para esse subespaço é:

$$b=(\begin{bmatrix}
0 &0\\
1 &0
\end{bmatrix})$$



## 3)

### a)

A equação é linear porque ela é expressa como potências de primeiro grau de derivadas de $x$, ou seja, os termos são lineares com respeito as derivadas de x.

### b)

In [70]:
import sympy as sp

x,t,a,b,c = sp.symbols('x,t,a , b ,c')

In [77]:
X=a*(t)+b

In [79]:
sp.diff(X,t)-3*(t**2)*X

a - 3*t**2*(a*t + b)

Logo,
$$(-3a)t^{3}+(-3b)t^{2}+a \equiv 1 -3t^{2}$$

$$\Rightarrow a=1 , b=0$$

1) está provada